src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,154 @@
+
+            // If the client has already asked for diagnostics for this document, see if we have actually recorded any
+            // differences, or if they should just use the same diagnostics as before.
+            if (request.PreviousResultId != null)
+            {
+                lock (_gate)
+                {
+                    if (_documentIdToLastResultId.TryGetValue((workspace, document.Id), out var lastReportedResultId) &&
+                        lastReportedResultId == request.PreviousResultId)
+                    {
+                        // Nothing changed between the last request and this one.  Report a null response to the client
+                        // to know they don't need to do anything.
+                        return null;
+                    }
+                }
+            }
+
+            // Being asked about this document for the first time.  Or being asked again and we have different diagnostics.
+
+            var diagnosticKeys = _diagnosticService.GetDiagnosticsUpdatedEventArgs(workspace, project.Id, document.Id, cancellationToken).ToImmutableArray();
+
+            using var _ = ArrayBuilder<LspDiagnostic>.GetInstance(out var result);
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            foreach (var key in diagnosticKeys)
+            {
+                var id = key.Id;
+                var documentId = key.DocumentId;
+
+                // ignore diagnostics unrelated to an actual document.
+                if (documentId == null)
+                    continue;
+
+                var diagnostics = _diagnosticService.GetDiagnostics(key.Workspace, key.ProjectId, key.DocumentId, key.Id, includeSuppressedDiagnostics: false, cancellationToken);
+                if (diagnostics == null)
+                    continue;
+
+                foreach (var diagnostic in diagnostics)
+                    result.AddIfNotNull(Convert(text, diagnostic));
+            }
+
+            string resultId;
+            lock (_gate)
+            {
+                // Keep track of the diagnostics we reported here so that we can short-circuit producing diagnostics for
+                // the same diagnostic set in the future.
+                resultId = _nextResultId++.ToString();
+                _documentIdToLastResultId[(workspace, document.Id)] = resultId;
+            }
+
+            return new[] { new DiagnosticReport { ResultId = resultId, Diagnostics = result.ToArray() } };
+        }
+
+        private static LspDiagnostic? Convert(SourceText text, DiagnosticData diagnosticData)
+        {
+            Contract.ThrowIfNull(diagnosticData.Message, $"Got a dicument diagnostic that did not haev a {nameof(diagnosticData.Message)}");
+            Contract.ThrowIfNull(diagnosticData.DataLocation, $"Got a dicument diagnostic that did not haev a {nameof(diagnosticData.DataLocation)}");
+
+            return new LspDiagnostic
```

Comment:
Can be done in a followup - but we should use the VSDiagnostic type
https://devdiv.visualstudio.com/DevDiv/_git/VSLanguageServerClient?path=%2Fsrc%2Fproduct%2FLanguageServer.Protocol.Extensions%2FDiagnostics%2FVSDiagnostic.cs&_a=contents&version=GBdevelop

Has support for way more tags, and we can fix our handling of hidden diagnostics (currently they are squiggled)

Reply:
i'll do that now.